compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/AbstractParser.java

```diff
@@ -226,12 +226,29 @@ protected void updateFirstNodeInListWithInvalidNode(List<STNode> nodeList,
                                                         STNode invalidParam,
                                                         DiagnosticCode diagnosticCode,
                                                         Object... args) {
-        STNode firstNode = nodeList.remove(0);
-        STNode newNode = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(firstNode, invalidParam);
+        updateANodeInListWithInvalidNode(nodeList, 0, invalidParam, diagnosticCode, args);
+    }
+
+    /**
+     * Clones the a node in list with the invalid node as leading minutiae and update the list.
+     *
+     * @param nodeList       node list to be updated
+     * @param indexOfTheNode index of the node in list to be updated
+     * @param invalidParam   the invalid node to be attached to the first node in list as minutiae
+     * @param diagnosticCode diagnostic code related to the invalid node
+     * @param args           additional arguments used in diagnostic message
+     */
+    protected void updateANodeInListWithInvalidNode(List<STNode> nodeList,
+                                                    int indexOfTheNode,
+                                                    STNode invalidParam,
+                                                    DiagnosticCode diagnosticCode,
+                                                    Object... args) {
+        STNode node = nodeList.remove(indexOfTheNode);
```

Comment:
Do we need to remove this? Can we use `get()` to retrieve and then `set()` to replace the original item?

Remove is a very expensive operation. It shifts all the trailing items by one. Lets try to minimize the usage of `remove()` in all the places